### DISCUSSION TITLE
Applying watermark to an animated file

### DISCUSSION THREAD

**Discussion by johntrunc** [libvips/libvips#2082] [Q&A] [answered]
Hi,

I need to apply a watermark to an animated image (GIF for example, but probably need the same for animated WebP) so it's showed in every page/frame. For non animated images I'm using `composite2` function but I'm wondering what's the best way to do that for an animated image.

I've been thinking into extract every page/frame into a vips image, apply the watermark with `composite2` and the join them with `arrayjoin` function but not sure if it's the best approach or if there is another common way to do this with libvips.

Thanks in advance.

**Comment by jcupitt** [comment]
Hi @johntrunc, I would make an overlay the same size as the frame, then use `replicate` to make n-pages vertical copies. You can then watermark every page in one operation.

**Comment by johntrunc** [comment]
Thanks @jcupitt I will explore that solution.

**Comment by johntrunc** [comment]
@johntrunc I have explored your suggested solution and it's working fine except for one scenario.

If I use `vips_thumbnail_image` to change the size of the image before applying the watermark everything is fine. I'm using `vips_image_get_page_height` and the returned value correspond to the correct value after the thumbnail operation.

But the same is not happening when I use `vips_resize`. If after a `vips_resize` I execute a call to `vips_image_get_page_height` I get the original page height of the image (not the correct value after the resize). Is that expected behavior?

Thank you very much.

**Comment by jcupitt** [comment]
That's expected behaviour -- it's up to you to update page height on crop or resize.

**Comment by johntrunc** [comment]
@jcupitt I need to do some operations that require process every frame individually.

I'm thinking into some process that will load frame by frame and process them one by one. What's the recommended way to do this? There is any common way to split the frames/pages into different images or I need to load the images one by one specifying a different value for the `page` parameter?

Thanks.

**Comment by jcupitt** [comment]
Hi again @johntrunc,

It'd be best to make a tiny benchmark in something like python and experiment, but usually you'd want to crop to a set of frames, process each one, then recombine. Something like (untested):

```python
original = pyvips.Image.new_from_file("big_animation.gif", n=-1, access="sequential")
page_height = original.get("page-height")
n_pages = original.get("n-pages")
pages = [original.crop(0, page_number * page_height, original.width, page_height) for page_number in range(0, n_pages)]
for i in range(len(pages)):
    pages[i] += 12 # some kind of processing
recombined = pyvips.Image.arrayjoin(pages, across=1)
# update page-height if necessary
recombined.write_to_file("new_animation.gif")
```

libvips should be able to evaluate that pipeline pretty efficiently. It'll work for multi-page formats like PDF as well.

**Comment by johntrunc** [comment]
Thanks John, I will try with your solution. Very much appreciated.

**Comment by johntrunc** [comment]
@jcupitt everything is working with your solution specifically for WebP but when try to use GIF as output all I get is a white image. This is only happening when I try to put another image (the watermark) over the original one (with `composite2`) and that image has 4 bands with transparency.

Something special is needed to be done for GIFs?

**Comment by jcupitt** [comment]
No, it ought to work, I think. Post an example that fails and I'll have a look.

**Comment by johntrunc** [comment]
I have several examples where this is happening. I will try with one that is simple. I'm using Go and my own integration with libvips.

In the following example I'm trying to add a border to the GIF, this is working for WebP without issues but I think it's not working for GIFs for some difference with the alpha channel I guess.

```go
        // 4 is the number of bands and params.BorderColor the specified BorderColor
	borderImg, err := vips.NewFromImage(img, 4, params.BorderColor.ToArrayDoubleRGBA())
	if err != nil {
		return err
	}
	defer borderImg.Close()

	borderSize := params.Border * 2
	width := img.GetWidth() - borderSize
	height := img.GetHeight() - borderSize

	err = vips.Run(
		borderImg,
                // The true at the end is the fill parameter
		vips.DrawRect(vips.ToArrayDouble(0.0, 0.0, 0.0, 0.0), params.Border, params.Border, width, height, true),
	)
	if err != nil {
		return err
	}

	return vips.Run(img, vips.Composite2(borderImg, vips.BlendModeOver, 0, 0))
```

I'm trying here with `borderImg` to create a new image with the same size and width that the original one (the GIF loaded in `img` that btw it's only one page of the GIF file) and filled with the color that we want. After that I'm drawing at the center a transparent rect so the original image is seen.

With this at the end I'm getting a totally white image. Tell me if I can give you some other examples and thank you very much for your help.

**Comment by johntrunc** [comment]
@jcupitt if it's necessary I can create a C example to clarify what is happening. Tell me if that can help you understanding the problem. Thank you very much.

**Comment by jcupitt** [comment]
Hi again @johntrunc,

I made a complete py version:

```python
#!/usr/bin/python3

import sys
import pyvips

# load the animation, chop into pages
image = pyvips.Image.new_from_file(sys.argv[1], n=-1, access="sequential")
page_height = image.get("page-height")
n_pages = image.get("n-pages")
pages = [image.crop(0, page_number * page_height, image.width, page_height)
         for page_number in range(0, n_pages)]

# load the overlay
overlay = pyvips.Image.new_from_file(sys.argv[3])

# render the overlay on every page
for i in range(len(pages)):
    pages[i] = pages[i].composite(overlay, 'over')

# recombine pages, save
recombined = pyvips.Image.arrayjoin(pages, across=1)
recombined.write_to_file(sys.argv[2])
```

And it seems to work for GIF and WEBP.

Could the problem be in `draw_rect`? It needs handling with a little care, since it modifies the argument image.

**Comment by johntrunc** [comment]
@jcupitt I think the problem (and the difference with your Python script) is that I'm generating a 4 band image to generate the border/overlay and then composing that 4 band image over the GIF. That's causing me to don't see the GIF image at all for some reason I don't know.

To generate the effect or a border with some specific width I'm generating an image with the same size of the GIF source page with 4 bands and filled with a specific color (I'm using red for example):

```go
// 4 is the number of bands and params.BorderColor the specified BorderColor
borderImg, err := vips.NewFromImage(img, 4, params.BorderColor.ToArrayDoubleRGBA())
```

Then I'm drawing a transparent rect in the center to allow the gif be showed after the compose:

```go
	err = vips.Run(
		borderImg,
                // The true at the end is the fill parameter
                // The Double array is the transparent color to fill the rect
		vips.DrawRect(vips.ToArrayDouble(0.0, 0.0, 0.0, 0.0), params.Border, params.Border, width, height, true),
	)
```

And using composite2 to put the border image over the gif:

```go
vips.Run(img, vips.Composite2(borderImg, vips.BlendModeOver, 0, 0))
```
(I'm joining the pages at the end too with `vips_arrayjoin`).

But at the end what I see is the following:

![example](https://user-images.githubusercontent.com/70328146/104036702-042e6c80-51d4-11eb-98d9-c083b2a9eff1.gif)

What I get is that white box in the center instead of the expected GIF. This issue is not happening saving the file as WebP and I think it's related with how transparent pixels are managed by GIF format once that it's saved or something similar.

If I instead of GIF I save the final file as WebP I correctly get the following (zipped because is not supported by Github):

[example.webp.zip](https://github.com/libvips/libvips/files/5788186/example.webp.zip)

I think if you try to recreate the same behavior in Python you should get the same result and understand the problem.


Thank you. I really appreciate your help.

**Comment by johntrunc** [comment]
@jcupitt did you have the chance to take a look to my last comment? I can create an example of this on C if necessary. Thanks for your work.

**Comment by jcupitt** [comment]
I had a quick look but didn't finish it before I was distracted by something else. I'll have another go.

**Comment by jcupitt** [comment]
This seems to work for me:

```python3
#!/usr/bin/python3

import sys
import pyvips

# load the animation, chop into pages
image = pyvips.Image.new_from_file(sys.argv[1], n=-1, access="sequential")
page_height = image.get("page-height")
n_pages = image.get("n-pages")
pages = [image.crop(0, page_number * page_height, image.width, page_height)
         for page_number in range(0, n_pages)]

# generate the overlay ... a solid red squarem, then draw a transparent rect 
# inside that
overlay = pages[0].new_from_image([255, 0, 0, 255])
overlay = overlay.draw_rect(0, 
                            10, 10, overlay.width - 20, overlay.height - 20,
                            fill=True)

# render the overlay on every page
for i in range(len(pages)):
    pages[i] = pages[i].composite(overlay, 'over')

# recombine pages, save
recombined = pyvips.Image.arrayjoin(pages, across=1)
recombined.write_to_file(sys.argv[2])
```

If I run:

```
./annotate-animated.py ~/pics/dancing_banana2.lossless.webp[n=-1] x.gif
```

I see:

![x](https://user-images.githubusercontent.com/580843/106016847-c8dadb80-60b7-11eb-8afa-2d6360079348.gif)

**Comment by johntrunc** [comment]
Hi @jcupitt

I have translated your Python script to C and is working fine for me too. After checking what I was doing different from the script I have found that the problem was in the `new_from_image` function. I was implementing something custom becase I didn't see that libvips has that function already.

After using `vips_image_new_from_image` instead of my custom implementation to create images from a previous one everything is working as expected.

Thank you very much for your help.

**Comment by johntrunc** [comment]
@jcupitt I have found again another place where I get a total white GIF output. I have created an small C program to show you what I'm doing.

I'm initially loading an animated WebP file, cropping every page of it. After that I'm loading a SVG and using `vips_linear` to apply some transparency to it. Finally I compose the SVG over every page, joining the pages and saving the output as GIF.

When I'm saving the file as WebP everything works as expected and you can see the SVG with some transparency over the animated input, but when I use GIF for the output file what I see is a completely white output.

```c
// gcc -g -Wall example.c `pkg-config vips --cflags --libs`
#include <stdio.h>
#include <vips/vips.h>

int
main(int argc, char **argv)
{
  if (VIPS_INIT("animations")) {
    vips_error_exit(NULL);
  }

  VipsImage *in, *svg, *out;

  if (!(in = vips_image_new_from_file("animation.webp", "n", -1, NULL))) {
    vips_error_exit(NULL);
  }

  int page_width = vips_image_get_width(in);
  int page_height = vips_image_get_page_height(in);
  int n_pages = vips_image_get_n_pages(in);

  VipsImage *pages[n_pages];

  for(int page_number = 0; page_number < n_pages; page_number++) {
    if (vips_crop(in, &pages[page_number], 0, page_number * page_height, page_width, page_height, NULL) == -1) {
      vips_error_exit(NULL);
    }
  }

  if (vips_svgload("example.svg", &svg, NULL) == -1) {
    vips_error_exit(NULL);
  }

  double a[] = {1.0, 1.0, 1.0, 0.5};
  double b[] = {0.0, 0.0, 0.0, 0.0};

  // Adding some transparency to the svg image
  if (vips_linear(svg, &svg, a, b, 4, NULL) == -1) {
    vips_error_exit(NULL);
  }

  for(int page_number = 0; page_number < n_pages; page_number++) {
    if (vips_composite2(pages[page_number], svg, &pages[page_number], VIPS_BLEND_MODE_OVER, NULL) == -1) {
      vips_error_exit(NULL);
    }
  }

  if (vips_arrayjoin(pages, &out, n_pages, "across", 1, NULL) == -1) {
    vips_error_exit(NULL);
  }

  // If I use here "output.webp" instead of "output.gif" everything works as expected
  // and we can see the svg partially transparent over the animated input file
  if (vips_image_write_to_file(out, "output.gif", NULL) == -1) {
    vips_error_exit(NULL);
  }

  return(0);
}
```
The two input files zipped:
- [animation.webp.zip](https://github.com/libvips/libvips/files/5947413/animation.webp.zip)
- [example.svg.zip](https://github.com/libvips/libvips/files/5947414/example.svg.zip)

The output that I'm getting:
- [output.webp.zip](https://github.com/libvips/libvips/files/5947493/output.webp.zip) (When saving as WebP the ouput it's correct).
- [output.gif.zip](https://github.com/libvips/libvips/files/5947514/output.gif.zip) (When saving as GIF the ouput is all white).

What I don't understand it's why this is not working for GIF but it's working when WebP is used?

As always thanks for your help.

**Reply by johntrunc** [reply]
@jcupitt did you have the chance to take a look to this last c example? Thanks for your time.

**Reply by jcupitt** [reply]
Hi, sorry for the delay, I should have some time this afternoon to take a look.

**Reply by johntrunc** [reply]
@jcupitt sorry for bother you again with this but by any chance did you take a look to it?. Thank you.

**Reply by johntrunc** [reply]
@jcupitt I have found a comment that could be related with my issue: https://github.com/libvips/libvips/issues/839#issuecomment-353725615

In that comment @kleisauke point out the following: 
> VIPS_FORMAT_FLOAT (range [-1, +1]) needs to be normalized to range [0, 1] (to make it compatible with write->storageType = FloatPixel).

And he adds a snippet that will output a wrong image if used. I have tried the same and I get a white GIF output in the same way that happens with my C example.

```sh
$ vips linear example.jpg test.v 1 200
$ vips magicksave test.v test.gif
```

I think these are related problems because I'm adding transparency with `vips_linear` to the svg file, so probably this is happening to me also.

**Comment by jcupitt** [comment, marked as answer]
Oh dear, sorry @johntrunc, this dropped off my radar again.

Try changing `linear` to be:

```
  if (vips_linear(svg, &svg, a, b, 4, "uchar", TRUE, NULL))
```

ie. make the output 8-bit (it'll be float by default). Here's a very slightly revised version of your code:

```C
// gcc -g -Wall johntrunc.c `pkg-config vips --cflags --libs`

#include <stdio.h>
#include <vips/vips.h>

int
main(int argc, char **argv)
{ 
  if (VIPS_INIT("animations")) 
    vips_error_exit(NULL);

  VipsImage *in, *svg, *out;
  
  if (argc != 4)
    vips_error_exit("usage: %s infile outfile overlay.svg", argv[0]);
  
  if (!(in = vips_image_new_from_file(argv[1], "n", -1, NULL)))
    vips_error_exit(NULL);
  
  int page_width = vips_image_get_width(in);
  int page_height = vips_image_get_page_height(in);
  int n_pages = vips_image_get_n_pages(in);
  
  VipsImage *pages[n_pages];
  
  for(int page_number = 0; page_number < n_pages; page_number++)
    if (vips_crop(in, &pages[page_number], 
      0, page_number * page_height, page_width, page_height, NULL))
      vips_error_exit(NULL);
  
  if (vips_svgload(argv[3], &svg, NULL))
    vips_error_exit(NULL);
  
  double a[] = {1.0, 1.0, 1.0, 0.5};
  double b[] = {0.0, 0.0, 0.0, 0.0};
  
  // Adding some transparency to the svg image
  if (vips_linear(svg, &svg, a, b, 4, "uchar", TRUE, NULL))
    vips_error_exit(NULL);
  
  for(int page_number = 0; page_number < n_pages; page_number++) 
    if (vips_composite2(pages[page_number], svg, 
      &pages[page_number], VIPS_BLEND_MODE_OVER, NULL)) 
      vips_error_exit(NULL);
  
  if (vips_arrayjoin(pages, &out, n_pages, "across", 1, NULL))
    vips_error_exit(NULL);

  if (vips_image_write_to_file(out, argv[2], NULL))    
    vips_error_exit(NULL);

  return(0);
}
```

If I run:

```
$ ./a.out ~/pics/dancing_banana2.lossless.gif x.gif ~/pics/lion.svg 
```

It generates:

![x](https://user-images.githubusercontent.com/580843/118012596-2504fb00-b349-11eb-89df-7592fc01efa9.gif)

libvips uses ImageMagick for GIF save and it looks like the magick GIF saver doesn't like float images.

**Reply by johntrunc** [reply]
Working perfectly using `uchar`.

As always thank you very much @jcupitt
